components/public-api/gitpod/experimental/v1/workspaces.proto

```diff
@@ -249,12 +251,37 @@ message WorkspaceInstanceStatus {
     // Admission describes who can access a workspace instance and its ports.
     AdmissionLevel admission = 6;
 
+    // ports is the list of served ports in the workspace.
+    repeated Port ports = 7;
+
     // repo details the Git working copy status of the workspace.
     // Note: this is a best-effort field and more often than not will not be present. Its absence does not
     // indicate the absence of a working copy.
     // contentservice.GitStatus repo = 7;
 }
 
+// PortVisibility defines who may access a workspace port which is guarded by an authentication in the proxy
+enum PortVisibility {
+    PORT_VISIBILITY_UNSPECIFIED = 0;
+
+    // Private means the port is accessible by the workspace owner only using the workspace port URL
+    PORT_VISIBILITY_PRIVATE = 1;
+
+    // Public means the port is accessible by everybody using the workspace port URL
+    PORT_VISIBILITY_PUBLIC = 2;
+}
+
+message Port {
+    // port number
+    uint64 port = 1;
+
+    // visiblity of this port
+    PortVisibility visibility = 2;
+
+    // url is the outward-facing URL where the port can be accessed on.
+    string url = 3;
+}
```

Comment:
Perhaps it makes sense to have the following objects to structure the request & response:

```
message PortSpec {
	 // port number
    uint64 port = 1;

    // visiblity of this port
    PortVisibility visibility = 2;
}

message PortStatus {
	// the URL which can be used to access the port
	string url = 1;
}
```

They would then be used in the following:
```
message Workspace {
	...
	repeated Port ports = 10;
}

message WorkspaceStatus {
	repeated PortStatus ports_status = 10;
}
```

Reply:
I will keep it first, and update it in the future if needed